### PR TITLE
fix: Fix type inference not working for new Try trait

### PR DESCRIPTION
**Pull request by Veykril**
Fixes https://github.com/rust-analyzer/rust-analyzer/issues/8907
bors r+

### CODE REVIEW

File: crates/hir_ty/src/infer.rs [hunk 1]
```diff
@@ -580,7 +580,10 @@ impl<'a> InferenceContext<'a> {
     fn resolve_ops_try_ok(&self) -> Option<TypeAliasId> {
         let path = path![core::ops::Try];
         let trait_ = self.resolver.resolve_known_trait(self.db.upcast(), &path)?;
-        self.db.trait_data(trait_).associated_type_by_name(&name![Ok])
+        let trait_data = self.db.trait_data(trait_);
+        trait_data
+            .associated_type_by_name(&name![Ok])
+            .or_else(|| trait_data.associated_type_by_name(&name![Output]))
```
Comment (lnicola) [inline crates/hir_ty/src/infer.rs:586]: Maybe add a `FIXME` here to remove the fallback?

**Comment by cynecx** [comment]
This doesn't actually fix https://github.com/rust-analyzer/rust-analyzer/issues/8907.

image

**Comment by Veykril** [comment]
Hmm, are we making more assumptions about the try trait somewhere? 🤔

**Comment by cynecx** [comment]
@Veykril It seems like trait solving fails because the `FromResidual` impl for `Result<T, F>` isn't included somehow?

**Comment by Veykril** [comment]
Heh, so using 
```rust
impl<T, E> crate::ops::FromResidual<Result<Infallible, E>> for Result<T, E> {}
```
fails, but using the default
```rust
impl<T, E> crate::ops::FromResidual for Result<T, E> {}
```
works

**Comment by cynecx** [comment]
ping @flodiebold

**Comment by flodiebold** [comment]
This seems to be a Chalk bug. It works without the explicit type parameter because that results in a slightly different cycle.

**Pull request merged**